File: Problem-1.cpp
/* ***************************
    1. Perform binary search and find the first occurance of target element from the left.
    2. If the 1st binary search does not return index of the target element return -1.
    3. Perform second binary search and look for the next occurance which appears on the end of the vector
    Time Complexity : O (log(N))
    Space Complexity : O(1)
    Did this code successfully run on Leetcode : Yes

 ****************************/

class Solution
{
public:
    vector<int> searchRange(vector<int> &nums, int target)
    {
        if (nums.empty())
        {
            vector<int> res = {-1, -1};
            return res;
        }

        int first = firstBinarySearch(nums, target);
        if (first == -1)
        {
            vector<int> res = {-1, -1};
            return res;
        }

        int second = secondBinarySearch(nums, target);

        vector<int> res = {first, second};
        return res;
    }

    int firstBinarySearch(vector<int> nums, int target)
    {
        int l = 0, h = nums.size() - 1;

        while (l <= h)
        {
            int mid = l + (h - l) / 2;
            if (nums[mid] == target)
            {
                if (mid == l || nums[mid - 1] != nums[mid])
                {
                    return mid;
                }
                h = mid - 1;
            }
            else if (nums[mid] < target)
            {
                l = mid + 1;
            }
            else
            {
                h = mid - 1;
            }
        }

        return -1;
    }

    int secondBinarySearch(vector<int> nums, int target)
    {
        int l = 0, h = nums.size() - 1;

        while (l <= h)
        {
            int mid = l + (h - l) / 2;
            if (nums[mid] == target)
            {
                if (mid == h || nums[mid + 1] != nums[mid])
                {
                    return mid;
                }
                l = mid + 1;
            }
            else if (nums[mid] < target)
            {
                l = mid + 1;
            }
            else
            {
                h = mid - 1;
            }
        }
        return -1;
    }
};

File: Problem_1.js
// Time Complexity : O(logn) + O(log n) +  (log n) i.e O(log n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach in three sentences only
/* Reused the binary search logic to first find the mid. If found, then searched the lowest and highest range of the target value
* by traversing half of the array to find lowest range first and then the other half to find higher range
*/

/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var searchRange = function(nums, target) {
    let range = [-1, -1];
    let low = 0;
    let high = nums.length-1;
    let targetIdx = -1;
    while(low <= high){
        let mid = low + Math.floor(high-low/2); //To avoid int overflow; 
        if(nums[mid] === target){
            targetIdx = mid;
            break;
        }
        else if(target < nums[mid]){
            high = mid -1;
        }
        else {
            low = mid + 1;
        }
    }
    if(targetIdx > -1 ){
        let lRange = targetIdx;
        let rRange = targetIdx;

        while(nums[lRange] === target && lRange >=0){
            lRange--;
        }
        while(nums[rRange] === target && rRange <=nums.length-1){
            rRange++;
        }
        return[lRange+1, rRange-1];

    }
    return [-1, -1];
};